presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java

```diff
@@ -0,0 +1,274 @@
+    }
+
+    public boolean isEmpty()
+    {
+        return totalRows == 0;
+    }
+
+    public List<Column> getColumns()
+    {
+        return columns;
+    }
+
+    public QueryResultRows getSelfOrNull()
+    {
+        if (isEmpty()) {
+            return null;
+        }
+
+        return this;
+    }
+
+    /**
+     * Returns expected row count (we don't know yet if every row is serializable)
+     *
+     * @return int
+     */
+    public int getTotalRows()
+    {
+        return totalRows;
+    }
+
+    public Long getUpdateCount()
+    {
+        if (hasUpdateCount()) {
+            requireNonNull(currentPage, "currentPage is null");
+
+            Block block = currentPage.getBlock(0);
+            Number value = (Number) types.get(0).getObjectValue(session, block, 0);
+
+            if (value != null) {
+                return value.longValue();
+            }
+        }
+
+        return null;
+    }
+
+    public boolean hasUpdateCount()
+    {
+        return !isEmpty() && columns.size() == 1 && columns.get(0).getType().equals(StandardTypes.BIGINT);
+    }
+
+    @Override
+    protected List<Object> computeNext()
+    {
+        loop:
+        while (true) {
+            inPageIndex++;
+
+            if (currentPage == null || (pages.isEmpty() && inPageIndex >= currentPage.getPositionCount())) {
+                return endOfData();
+            }
+            else if (inPageIndex >= currentPage.getPositionCount()) {
+                currentPage = pages.pollFirst();
+                requireNonNull(currentPage, "currentPage is null");
+                inPageIndex = 0;
+            }
+
+            rowPosition++;
+            List<Object> values = new ArrayList<>(columns.size());
+
+            for (int channel = 0; channel < currentPage.getChannelCount(); channel++) {
+                Type type = types.get(channel);
+                Block block = currentPage.getBlock(channel);
+
+                try {
+                    values.add(channel, type.getObjectValue(session, block, inPageIndex));
+                }
+                catch (Throwable throwable) {
+                    storeLastException(rowPosition, channel, throwable);
+                    // if row contains unserializable value - skip row entirely
+                    continue loop;
+                }
+            }
+
+            return unmodifiableList(values);
+        }
+    }
+
+    private void storeLastException(final int row, final int column, final Throwable cause)
+    {
+        // columns and rows are 0-indexed
+        PrestoException exception = new PrestoException(SERIALIZATION_ERROR,
+                format("Could not serialize type '%s' value at position %d:%d", columns.get(column).getType(), row + 1, column + 1),
+                cause);
+
+        lastException = exception;
+
+        if (exceptionConsumer != null) {
+            exceptionConsumer.accept(exception);
+        }
+    }
+
+    public Optional<Throwable> getLastSerializationException()
```

Comment:
This "last exception" seems to only be used for testing. Let's remove it, since we can and already do use the consumer for testing.